zeebe/broker-client/src/main/java/io/camunda/zeebe/broker/client/impl/RoundRobinDispatchStrategy.java

```diff
@@ -10,30 +10,98 @@
 import io.camunda.zeebe.broker.client.api.BrokerClusterState;
 import io.camunda.zeebe.broker.client.api.BrokerTopologyManager;
 import io.camunda.zeebe.broker.client.api.RequestDispatchStrategy;
+import io.camunda.zeebe.dynamic.config.state.RoutingState;
+import java.util.Set;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * Return the next partition using a round-robin strategy, but skips the partitions where there is
  * no leader at the moment.
  */
 public final class RoundRobinDispatchStrategy implements RequestDispatchStrategy {
 
-  private final AtomicInteger partitions = new AtomicInteger(0);
+  /** Holds the current partition ring. Starts off uninitialized and is updated on every request. */
+  private final AtomicReference<VersionedPartitionRing> partitionRing =
+      new AtomicReference<>(VersionedPartitionRing.uninitialized());
+
+  private final AtomicInteger offset = new AtomicInteger(0);
 
   @Override
   public int determinePartition(final BrokerTopologyManager topologyManager) {
     final BrokerClusterState topology = topologyManager.getTopology();
 
-    if (topology != null) {
-      for (int i = 0; i < topology.getPartitionsCount(); i++) {
-        final int offset = partitions.getAndIncrement();
-        final int partition = topology.getPartition(offset);
-        if (topology.getLeaderForPartition(partition) != BrokerClusterState.NODE_ID_NULL) {
-          return partition;
-        }
+    if (topology == null) {
+      return BrokerClusterState.PARTITION_ID_NULL;
+    }
+
+    final var partitions = updatePartitionRing(topologyManager);
```

Comment:
Good idea :+1: 
However, I don't see a neat way to implement this. We'd have to add another `AtomicInteger` to track at which offset we've last updated the partition ring. That probably has the same overhead as looking up the routing state version and comparing that to the current `VersionedPartitionRing`.

Reply:
I think we do something similar in the cluster gossip and swim by keeping a modifiable list. When a partition is used, it will be removed from the list. When the list is empty, update it from the partition ring. But ya, may be it is not worth the effort.